new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,109 @@
+import {
+    createContext,
+    useEffect,
+    useState,
+} from "react";
+
+
+interface UrlContextType {
+    setSearchParamSet: (searchParamSet: Record<string, string | null>) => void;
+    setHashParamSet: (hashParamSet: Record<string, string | null>) => void;
+    copyToClipboard: (
+        searchParamSet: Record<string, string | null>,
+        hashParamSet: Record<string, string | null>
+    ) => void;
+}
+
+const UrlContext = createContext <UrlContextType>({} as UrlContextType);
+
+
+interface UrlContextProviderProps {
+    children: React.ReactNode
+}
+
+/**
+ * Provides a context for managing URL parameters and hash values,
+ * including utilities for setting search and hash parameters,
+ * and copying the current URL with these parameters to the clipboard.
+ *
+ * @param children.children
+ * @param children The child components that will have access to the context.
+ */
+const UrlContextProvider = ({children}: UrlContextProviderProps) => {
+    const [hashParam, setHashParam] = useState<string>(window.location.hash.substring(1));
+    useEffect(() => {
+        setHashParam(window.location.hash.substring(1));
+    }, [window.location.hash]);
+
+    const setSearchParamSetHelper = (searchParamSet: Record<string, string | null>) => {
+        const newSearchParam = new URLSearchParams(window.location.search.substring(1));
+        const {filePath} = searchParamSet;
+        delete searchParamSet.filePath;
+
+        for (const [key, value] of Object.entries(searchParamSet)) {
+            if (null === value) {
+                newSearchParam.delete(key);
+            } else {
+                newSearchParam.set(key, value);
+            }
+        }
+        if (filePath) {
+            newSearchParam.set("filePath", filePath);
+        }
+
+        return newSearchParam;
+    };
+
+    const setSearchParamSet = (searchParamSet: Record<string, string | null>) => {
+        const newUrl = new URL(window.location.href);
+        newUrl.search = setSearchParamSetHelper(searchParamSet).toString();
+        if (!(/%23|%26/).test(newUrl.search)) {
+            newUrl.search = decodeURIComponent(newUrl.search);
+        }
+        window.history.pushState({}, "", newUrl.toString());
+    };
+
+    const setHashParamSetHelper = (hashParamSet: Record<string, string | null>) => {
```

Comment:
Let's avoid naming helper functions as "helpers". Try to come up with names that are more descriptive.

```suggestion
    const getUpdatedSearchParams= (newSearchParams: Record<string, string | null>) => {
```

Reply:
Since we're able to define types in TypeScript, we can avoid including types in the variable names.

Reply:
> Since we're able to define types in TypeScript, we can avoid including types in the variable names.

I don't quite get it. Do you mean for `hashParamSet`, we don't need `Record<string, string | null>`? But wouldn't that give an implicit type of any?

Reply:
Sorry, I meant we can name this parameter as `hashParams` since the `set` in the name is a bit confusing. Strictly speaking, the parameter is also a map (of key-value pairs) rather than a set (of unique values), right?

To make the purpose of the parameter more clear, I renamed the param as `updates` in the patch provided. Please start the modifications with the patch.